In [2]:
#import the necessary library

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pandas.plotting import scatter_matrix
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [10]:
#load the data set
data=pd.read_csv('Life Expectancy Data.csv')
data.head()

Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   ...  \
0             62     0.01               71.279624         65.0      1154  ...   
1             64     0.01               73.523582         62.0       492  ...   
2             66     0.01               73.219243         64.0       430  ...   
3             69     0.01               78.184215         67.0      2787  ...   
4             71     0.01                7.097109         68.0      3013  ...   

   Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0    6.0               8.16         65.0        0.1  584.259210  33736494.0   
1   58.0               8.18         62.0        0.1  612.696514    327582.0   
2   62.0               8.13         64.0        0.1  631.744976  31731688.0   
3   67.0               8.52         67.0        0.1  669.959000   3696958.0   
4   68.0               7.87         68.0        0.1   63.537231   2978599.0   

    thinness  1-19 years   thinness 5-9 years  \
0                   17.2                 17.3   
1                   17.5                 17.5   
2                   17.7                 17.7   
3                   17.9                 18.0   
4                   18.2                 18.2   

   Income composition of resources  Schooling  
0                            0.479       10.1  
1                            0.476       10.0  
2                            0.470        9.9  
3                            0.463        9.8  
4                            0.454        9.5  

[5 rows x 22 columns]

In [11]:
# prompt: convert Country and Status using label encoding

from sklearn.preprocessing import LabelEncoder

# Assuming 'data' DataFrame is already loaded as in the provided code

# Create a LabelEncoder object
label_encoder = LabelEncoder()

# Fit and transform the 'Country' column
data['Country'] = label_encoder.fit_transform(data['Country'])

# Fit and transform the 'Status' column
data['Status'] = label_encoder.fit_transform(data['Status'])

# Display the first few rows to verify the changes
data.head()

Country  Year  Status  Life expectancy   Adult Mortality  infant deaths  \
0        0  2015       1              65.0            263.0             62   
1        0  2014       1              59.9            271.0             64   
2        0  2013       1              59.9            268.0             66   
3        0  2012       1              59.5            272.0             69   
4        0  2011       1              59.2            275.0             71   

   Alcohol  percentage expenditure  Hepatitis B  Measles   ...  Polio  \
0     0.01               71.279624         65.0      1154  ...    6.0   
1     0.01               73.523582         62.0       492  ...   58.0   
2     0.01               73.219243         64.0       430  ...   62.0   
3     0.01               78.184215         67.0      2787  ...   67.0   
4     0.01                7.097109         68.0      3013  ...   68.0   

   Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0               8.16         65.0        0.1  584.259210  33736494.0   
1               8.18         62.0        0.1  612.696514    327582.0   
2               8.13         64.0        0.1  631.744976  31731688.0   
3               8.52         67.0        0.1  669.959000   3696958.0   
4               7.87         68.0        0.1   63.537231   2978599.0   

    thinness  1-19 years   thinness 5-9 years  \
0                   17.2                 17.3   
1                   17.5                 17.5   
2                   17.7                 17.7   
3                   17.9                 18.0   
4                   18.2                 18.2   

   Income composition of resources  Schooling  
0                            0.479       10.1  
1                            0.476       10.0  
2                            0.470        9.9  
3                            0.463        9.8  
4                            0.454        9.5  

[5 rows x 22 columns]

In [12]:
#checking for any null or empty cells in data
data.isna().sum()

Country                              0
Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64

In [13]:
# prompt: remove all the null values from all the columns

# Drop rows with any null values
data.dropna(inplace=True)

# Display the first few rows to verify the changes
data.head()

# Check for null values again to confirm removal
data.isna().sum()

Country                            0
Year                               0
Status                             0
Life expectancy                    0
Adult Mortality                    0
infant deaths                      0
Alcohol                            0
percentage expenditure             0
Hepatitis B                        0
Measles                            0
 BMI                               0
under-five deaths                  0
Polio                              0
Total expenditure                  0
Diphtheria                         0
 HIV/AIDS                          0
GDP                                0
Population                         0
 thinness  1-19 years              0
 thinness 5-9 years                0
Income composition of resources    0
Schooling                          0
dtype: int64

In [14]:
#checking for datatype of columns
data.dtypes

Country                              int64
Year                                 int64
Status                               int64
Life expectancy                    float64
Adult Mortality                    float64
infant deaths                        int64
Alcohol                            float64
percentage expenditure             float64
Hepatitis B                        float64
Measles                              int64
 BMI                               float64
under-five deaths                    int64
Polio                              float64
Total expenditure                  float64
Diphtheria                         float64
 HIV/AIDS                          float64
GDP                                float64
Population                         float64
 thinness  1-19 years              float64
 thinness 5-9 years                float64
Income composition of resources    float64
Schooling                          float64
dtype: object

In [17]:
#assign dependent and independent variables
X = data[[" BMI "]]
y = data["Life expectancy "]

In [18]:
# Spliting the data set for training and testing in random way
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [19]:
#assiging LinearRegression() method to variable for reusability
lr = LinearRegression()

In [20]:
#pass training dataset into the assigned LinearRegression() method
lr.fit(X_train.values,y_train)

LinearRegression()

In [21]:
#predict y values based on x column test data
y_pred = lr.predict(X_test)

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:486: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


In [22]:
#print r-sqaure value
round(r2_score( y_test, y_pred),2)

0.32

In [23]:
#print intercept score
print(round(lr.intercept_,2))

60.07


In [24]:
#checking the prediction
lr.predict(X[0:1])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:486: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


array([64.64016933])

In [25]:
# to check datatype of linear regression
type(lr)

sklearn.linear_model._base.LinearRegression

In [26]:
#serialization
#model persistance (saving and loading trained models)
import pickle

In [27]:
# save: 'with' is package deals with file handeling, wb- write
with open('model.pkl','wb') as f:
    pickle.dump(lr,f)

In [28]:
# load: rb-read
with open('model.pkl', 'rb') as f:
    clf2 = pickle.load(f)

In [29]:
#predict value of y for 1st 5 rows of x
clf2.predict(X[0:5])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:486: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


array([64.64016933, 64.52055152, 64.4009337 , 64.28131588, 64.18562162])

In [30]:
# Display the first few predictions alongside the actual values
predictions = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})
print(predictions.head())

      Actual  Predicted
1210    67.5  60.668858
2273    73.8  73.252652
1005    79.1  73.731124
2926    54.9  67.223914
518     48.6  64.807634


# **Interpretations Simple Linear Regression Analysis**

**Model Overview**: Simple Linear Regression model performed to evaluate the relationship between BMI (independent variable) and Life expectancy (dependent variable). The dataset was split into training and testing sets, with 80% of data used for training the model and 20% reserved for testing.

**Model Parameters:**

**a) Intercept (𝛽0) = 60.07**, which indicates that if BMI is 0, the predicted Profit would be  60.07 units. While BMI cannot realistically be 0, this value provides a starting point for the model's predictions.

**b) R-Squared = 0.32**, suggests that the model has limited explanatory power. Only 32% of the variability in life expectancy is explained by BMI. This indicates that other factors (e.g., healthcare access, education, economic conditions) play a significant role in determining life expectancy.